### ISSUE TITLE
In the draw_specific_kpts method of the SolutionAnnotator class, the order of line drawing is independent of the order of the input arguments.

### ISSUE THREAD

**Issue by cjt24703** [ultralytics/ultralytics#23652]
### Search before asking

- [x] I have searched the Ultralytics YOLO [issues](https://github.com/ultralytics/ultralytics/issues) and found no similar bug report.

### Bug

<img width="1037" height="602" alt="Image" src="https://github.com/user-attachments/assets/db531024-ad3d-4b59-ab47-19d882d5baa1" />

### Environment

Ultralytics 8.4.6 🚀 Python-3.11.14 torch-2.4.0+cu121 CPU (11th Gen Intel Core i7-11700 @ 2.50GHz)
Setup complete ✅ (16 CPUs, 31.1 GB RAM, 265.5/439.9 GB disk)

OS                     Linux-5.18.17-amd64-desktop-hwe-x86_64-with-glibc2.28
Environment            Linux
Python                 3.11.14
Install                pip
Path                   /home/guanghui/miniconda3/envs/unsloth_env/lib/python3.11/site-packages/ultralytics
RAM                    31.13 GB
Disk                   265.5/439.9 GB
CPU                    11th Gen Intel Core i7-11700 @ 2.50GHz
CPU count              16
GPU                    None
GPU count              None
CUDA                   None

numpy                  ✅ 1.26.4>=1.23.0
matplotlib             ✅ 3.10.8>=3.3.0
opencv-python          ✅ 4.11.0.86>=4.6.0
pillow                 ✅ 11.3.0>=7.1.2
pyyaml                 ✅ 6.0.3>=5.3.1
requests               ✅ 2.32.5>=2.23.0
scipy                  ✅ 1.16.3>=1.4.1
torch                  ✅ 2.4.0>=1.8.0
torch                  ✅ 2.4.0!=2.4.0,>=1.8.0; sys_platform == "win32"
torchvision            ✅ 0.19.0>=0.9.0
psutil                 ✅ 7.2.1>=5.8.0
polars                 ✅ 1.37.1>=0.20.0
ultralytics-thop       ✅ 2.0.18>=2.0.18

### Minimal Reproducible Example

from ultralytics import solutions
import cv2

cap = cv2.VideoCapture("./video1.mp4")
assert cap.isOpened(), "Error reading video file"
w, h, fps = (int(cap.get(x)) for x in (cv2.CAP_PROP_FRAME_WIDTH, cv2.CAP_PROP_FRAME_HEIGHT, cv2.CAP_PROP_FPS))

gym = solutions.AIGym(
    line_width=2,
    show=True,
    kpts= [8,6,12],
)

while cap.isOpened():
    success, im0 = cap.read()
    if not success:
        print("Video frame is empty or processing is complete.")
        break
    results = gym(im0)
    print(results)

cv2.destroyAllWindows()

### Additional

You can refer to the code fix.

def draw_specific_kpts(
        self,
        keypoints: list[list[float]],
        indices: list[int] | None = None,
        radius: int = 2,
        conf_thresh: float = 0.25,
    ) -> np.ndarray:
        """Draw specific keypoints for gym steps counting.

        Args:
            keypoints (list[list[float]]): Keypoints data to be plotted, each in format [x, y, confidence].
            indices (list[int], optional): Keypoint indices to be plotted.
            radius (int): Keypoint radius.
            conf_thresh (float): Confidence threshold for keypoints.

        Returns:
            (np.ndarray): Image with drawn keypoints.

        Notes:
            Keypoint format: [x, y] or [x, y, confidence].
            Modifies self.im in-place.
        """
        indices = indices or [2, 5, 7]
        points = [(int(k[0]), int(k[1])) for i, k in enumerate(keypoints) if i in indices and k[2] >= conf_thresh]
        # points = [((int(k[0]),int(k[1]))) for i in indices for i1,k in enumerate(keypoints) if i == i1 and k[2] >= conf_thresh]
        # Draw lines between consecutive points
        for start, end in zip(points[:-1], points[1:]):
            cv2.line(self.im, start, end, (0, 255, 0), 2, lineType=cv2.LINE_AA)

        # Draw circles for keypoints
        for pt in points:
            cv2.circle(self.im, pt, radius, (0, 0, 255), -1, lineType=cv2.LINE_AA)

        return self.im

### Are you willing to submit a PR?

- [x] Yes I'd like to help by submitting a PR!

**Comment by UltralyticsAssistant** [comment]
👋 Hello @cjt24703, thank you for your interest in Ultralytics 🚀! We recommend a visit to the [Docs](https://docs.ultralytics.com/) for new users where you can find many [Python](https://docs.ultralytics.com/usage/python/) and [CLI](https://docs.ultralytics.com/usage/cli/) usage examples and where many of the most common questions may already be answered.

This is an automated response 🤖—an Ultralytics engineer will also assist you soon. Thanks for the clear report and for offering to submit a PR 🙌

If this is a 🐛 Bug Report, please provide a [minimum reproducible example](https://docs.ultralytics.com/help/minimum-reproducible-example/) to help us debug it. (It looks like you included one—if you can, please make it fully self-contained by including a tiny sample input/video or a short snippet that reproduces the behavior without external files, plus the expected vs actual drawing order.)

If this is a custom training ❓ Question, please provide as much information as possible, including dataset image examples and training logs, and verify you are following our [Tips for Best Training Results](https://docs.ultralytics.com/guides/model-training-tips/).

Join the Ultralytics community where it suits you best. For real-time chat, head to [Discord](https://discord.com/invite/ultralytics) 🎧. Prefer in-depth discussions? Check out [Discourse](https://community.ultralytics.com/). Or dive into threads on our [Subreddit](https://www.reddit.com/r/Ultralytics/) to share knowledge with the community.

## Upgrade

Upgrade to the latest `ultralytics` package including all [requirements](https://github.com/ultralytics/ultralytics/blob/main/pyproject.toml) in a [**Python>=3.8**](https://www.python.org/) environment with [**PyTorch>=1.8**](https://pytorch.org/get-started/locally/) to verify your issue is not already resolved in the latest version:

```bash
pip install -U ultralytics
```

## Environments

YOLO may be run in any of the following up-to-date verified environments (with all dependencies including [CUDA](https://developer.nvidia.com/cuda)/[CUDNN](https://developer.nvidia.com/cudnn), [Python](https://www.python.org/) and [PyTorch](https://pytorch.org/) preinstalled):

- **Notebooks** with free GPU: <a href="https://console.paperspace.com/github/ultralytics/ultralytics"><img src="https://assets.paperspace.io/img/gradient-badge.svg" alt="Run on Gradient"/></a> <a href="https://colab.research.google.com/github/ultralytics/ultralytics/blob/main/examples/tutorial.ipynb"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"></a> <a href="https://www.kaggle.com/models/ultralytics/yolo26"><img src="https://kaggle.com/static/images/open-in-kaggle.svg" alt="Open In Kaggle"></a>
- **Google Cloud** Deep Learning VM. See [GCP Quickstart Guide](https://docs.ultralytics.com/yolov5/environments/google_cloud_quickstart_tutorial/)
- **Amazon** Deep Learning AMI. See [AWS Quickstart Guide](https://docs.ultralytics.com/yolov5/environments/aws_quickstart_tutorial/)
- **Docker Image**. See [Docker Quickstart Guide](https://docs.ultralytics.com/yolov5/environments/docker_image_quickstart_tutorial/) <a href="https://hub.docker.com/r/ultralytics/ultralytics"><img src="https://img.shields.io/docker/pulls/ultralytics/ultralytics?logo=docker" alt="Docker Pulls"></a>

## Status

<a href="https://github.com/ultralytics/ultralytics/actions/workflows/ci.yml?query=event%3Aschedule"><img src="https://github.com/ultralytics/ultralytics/actions/workflows/ci.yml/badge.svg" alt="Ultralytics CI"></a>

If this badge is green, all [Ultralytics CI](https://github.com/ultralytics/ultralytics/actions/workflows/ci.yml?query=event%3Aschedule) tests are currently passing. CI tests verify correct operation of all YOLO [Modes](https://docs.ultralytics.com/modes/) and [Tasks](https://docs.ultralytics.com/tasks/) on macOS, Windows, and Ubuntu every 24 hours and on every commit.

**Comment by glenn-jocher** [comment]
Good catch — `SolutionAnnotator.draw_specific_kpts()` currently builds `points` by enumerating all keypoints, so the line order follows the native keypoint index order, not the user-provided `indices` order (see the current implementation in the [draw_specific_kpts() reference](https://docs.ultralytics.com/reference/solutions/solutions/#ultralytics.solutions.solutions.SolutionAnnotator.draw_specific_kpts)); a PR is welcome to construct `points` by iterating `indices` directly (so `[8, 6, 12]` draws `8→6→12`), e.g.:  
```python
indices = indices or [2, 5, 7]
points = []
for j in indices:
    kp = keypoints[j]
    conf = float(kp[2]) if len(kp) > 2 else 1.0
    if conf >= conf_thresh:
        points.append((int(kp[0]), int(kp[1])))
```
and please include a small regression test and keep the change scoped per the [contributing guide](https://docs.ultralytics.com/help/contributing/).

**Comment by krgkaushik** [comment]
Hi @cjt24703, thanks for the PR!

I noticed the implementation uses enumerate(keypoints), which will always process points in ascending order (e.g., index 6, then 8, then 12) regardless of the order in the indices list.

To support custom connection paths (like connecting 8 → 6 → 12), we should iterate over the indices list directly, as Glenn suggested.

Suggested change:
``points = []
for i in indices:
    # Check if index is within bounds and confidence is high enough
    if i < len(keypoints) and keypoints[i][2] >= conf_thresh:
        points.append((int(keypoints[i][0]), int(keypoints[i][1])))``

**Comment by cjt24703** [comment]
Thank you all for your comments. I have made modifications based on Glenn's suggestions. This is my first time submitting an RP, so if there are any other issues, please let me know. I wish everyone a happy Lunar New Year

**Comment by glenn-jocher** [comment]
Thanks for the update, and happy Lunar New Year to you too — if your PR now iterates `indices` directly, please also add a small guard for out-of-range indices and for keypoints that may be `[x, y]` (no confidence) vs `[x, y, conf]` (see how keypoints can be 2D/3D in the [Keypoints results reference](https://docs.ultralytics.com/reference/engine/results/#ultralytics.engine.results.Keypoints)), and include a tiny regression test that confirms `indices=[8, 6, 12]` connects `8→6→12` as used in the `AIGym` workflow (see [AIGym.process reference](https://docs.ultralytics.com/reference/solutions/ai_gym/#ultralytics.solutions.ai_gym.AIGym.process)).

**Comment by mango766** [comment]
I've opened PR #23913 to address this — it iterates `indices` directly (preserving user-specified order like `[8, 6, 12]` → `8→6→12`), adds a bounds guard for out-of-range indices, handles both `[x, y]` and `[x, y, conf]` keypoint formats, and includes three regression tests as @glenn-jocher suggested. Happy to adjust if anything needs tweaking.

**Comment by glenn-jocher** [comment]
Thanks for opening `#23913` — the change set sounds aligned with the expected fix, and the regression tests help; we’ll continue any requested tweaks on the PR thread, so please keep any follow-up commits scoped to this bug only.

**Comment by UltralyticsAssistant** [comment]
A potential fix for this issue has now been merged in [PR #24099: `draw_specific_kpts` to respect user-specified indices order](https://github.com/ultralytics/ultralytics/pull/24099) 🎉

### What changed
This PR updates `draw_specific_kpts()` in `ultralytics/solutions/solutions.py` in a few ways that may resolve the reported behavior:

- Preserves the exact order of the user-provided `indices` list when drawing keypoints and connecting lines
- Safely ignores out-of-range indices instead of failing
- Improves handling for keypoints that may not include confidence values
- Clarifies in the docstring that drawing follows the provided `indices` order

These changes should make custom keypoint drawing more predictable and robust.

### Credits
Thanks to @onuralpszr for the fix, with contributions from @Laughing-q 🙌

### How to test
Since this has been merged, it can be tested immediately from the latest `main` branch with:

```bash
pip install git+https://github.com/ultralytics/ultralytics.git@main
```

Alternatively, it will also be available in the next official PyPI release if preferred.

### Feedback
If possible, please test the updated code and share whether it resolves the issue. Thanks 🙏 for reporting this, and any follow-up feedback is very welcome if the issue still persists.
